src/client/cli/commands/archive.js

```diff
@@ -0,0 +1,145 @@
+        outputStream = (output == null) ? process.stdout : fs.createWriteStream(output)
+        tarball.pipe(gzip).pipe(outputStream)
+
+        let stmtBatch: Array<string> = []
+        let stmtBatchNumber = 0
+        function writeStatementBatch (force: boolean = false) {
+          if (force || stmtBatch.length >= STMT_BATCH_SIZE) {
+            const content = Buffer.from(stmtBatch.join('\n'), 'utf-8')
+            const filename = `stmt/${leftpad(stmtBatchNumber.toString(), 8)}.ndjson`
+            writeToTarball(tarball, filename, content)
+            stmtBatchNumber += 1
+            stmtBatch = []
+          }
+        }
+
+        outputStream.on('error', err => {
+          reject(new Error(`Error writing to ${outStreamName}: ${err.message}`))
+        })
+
+        queryStream.on('error', err => {
+          reject(new Error(`Error reading from query result stream: ${err.message}`))
+        })
+
+        queryStream.on('data', obj => {
+          let stmt
+          try {
+            stmt = Statement.fromProtobuf(obj)
+          } catch (err) {
+            // ignore non-statement results
+            return
+          }
+
+          stmtBatch.push(JSON.stringify(obj))
+          writeStatementBatch()
+
+          for (const id of stmt.objectIds) {
+            objectIds.add(id)
+          }
+          for (const id of stmt.depsSet) {
+            objectIds.add(id)
+          }
+        })
+
+        queryStream.on('end', () => {
+          writeStatementBatch(true)
+          resolve()
+        })
+      }))
+      .then(() => writeDataObjectsToTarball(client, tarball, objectIds))
+      .then(() => new Promise(resolve => {
+        outputStream.on('end', () => resolve())
+        tarball.finalize()
+      }))
+  })
+}
+
+function writeToTarball (tarball: Object, filename: string, content: Buffer) {
+  const header = Object.assign({}, {name: filename, size: content.length}, TAR_ENTRY_OPTS)
+  tarball.entry(header, content)
+}
+
+function writeDataObjectsToTarball (client: RestClient, tarball: Object, objectIds: Set<string>): Promise<*> {
+  if (objectIds.size < 1) return Promise.resolve()
+
+  const batchPromises = []
+  let batch = []
+  for (const id of objectIds) {
+    batch.push(id)
+    if (batch.length >= OBJECT_BATCH_SIZE) {
+      const ids = batch
+      batch = []
+      batchPromises.push(fetchObjectBatch(client, tarball, ids))
+    }
+  }
+
+  batchPromises.push(fetchObjectBatch(client, tarball, batch))
+  return Promise.all(batchPromises)
+}
+
+function fetchObjectBatch (client: RestClient, tarball: Object, objectIds: Array<string>): Promise<*> {
+  if (objectIds.length < 1) return Promise.resolve()
+
+  return client.batchGetDataStream(objectIds, false)
+    .then(stream => new Promise((resolve, reject) => {
+      stream.on('data', dataResult => {
+        const key = objectIds.shift()
+        if (dataResult == null || typeof dataResult !== 'object' || dataResult.data == null) return
```

Comment:
so if an object i missing from the datastore (or some other error condition), we just silently stop adding objects and still produce an archive?
I think we should at least notify the user that there was an error and possibly keep going and add the remaining objects.